pkg/services/folder/folderimpl/sqlstore.go

```diff
@@ -342,27 +346,155 @@ func (ss *sqlStore) GetHeight(ctx context.Context, foldrUID string, orgID int64,
-		}
-		return sess.SQL(b.String(), args...).Find(&folders)
+		return batch(len(q.UIDs), int(q.BatchSize), func(start, end int) error {
+			partialFolders := make([]*folder.Folder, 0, q.BatchSize)
+			partialUIDs := q.UIDs[start:min(end, len(q.UIDs))]
+			s := strings.Builder{}
+			s.WriteString(`SELECT f0.id, f0.org_id, f0.uid, f0.parent_uid, f0.title, f0.description, f0.created, f0.updated`)
+			// compute full path column if requested
+			if q.WithFullpath {
+				s.WriteString(fmt.Sprintf(`, %s AS fullpath`, getFullpathSQL(ss.db.GetDialect())))
+			}
+			s.WriteString(` FROM folder f0`)
+			// join the same table multiple times to compute the full path of a folder
+			if q.WithFullpath || len(q.ancestorUIDs) > 0 {
+				s.WriteString(getFullpathJoinsSQL())
+			}
+			s.WriteString(` WHERE f0.org_id=?`)
+			args := []any{q.OrgID}
+			if len(partialUIDs) > 0 {
+				s.WriteString(` AND f0.uid IN (?` + strings.Repeat(", ?", len(partialUIDs)-1) + `)`)
+				for _, uid := range partialUIDs {
+					args = append(args, uid)
+				}
+			}
+
+			if len(q.ancestorUIDs) == 0 {
+				err := sess.SQL(s.String(), args...).Find(&partialFolders)
+				if err != nil {
+					return err
+				}
+				folders = append(folders, partialFolders...)
+				return nil
+			}
+
+			// filter out folders if they are not in the subtree of the given ancestor folders
+			if err := batch(len(q.ancestorUIDs), int(q.BatchSize), func(start2, end2 int) error {
+				s2, args2 := getAncestorsSQL(ss.db.GetDialect(), q.ancestorUIDs, start2, end2, s.String(), args)
+				err := sess.SQL(s2, args2...).Find(&partialFolders)
+				if err != nil {
+					return err
+				}
+				folders = append(folders, partialFolders...)
+				return nil
+			}); err != nil {
+				return err
+			}
+			return nil
+		})
 	}); err != nil {
 		return nil, err
 	}
 
 	// Add URLs
 	for i, f := range folders {
+		f.Fullpath = strings.TrimLeft(f.Fullpath, "/")
 		folders[i] = f.WithURL()
 	}
 
 	return folders, nil
 }
+
+func getFullpathSQL(dialect migrator.Dialect) string {
+	concatCols := make([]string, 0, folder.MaxNestedFolderDepth)
+	concatCols = append(concatCols, "COALESCE(REPLACE(f0.title, '/', '\\/'), '')")
+	for i := 1; i <= folder.MaxNestedFolderDepth; i++ {
+		concatCols = append([]string{fmt.Sprintf("COALESCE(REPLACE(f%d.title, '/', '\\/'), '')", i), "'/'"}, concatCols...)
+	}
+	return dialect.Concat(concatCols...)
+}
+
+// getFullpathJoinsSQL returns a SQL fragment that joins the same table multiple times to get the full path of a folder.
+func getFullpathJoinsSQL() string {
+	joins := make([]string, 0, folder.MaxNestedFolderDepth)
+	for i := 1; i <= folder.MaxNestedFolderDepth; i++ {
+		joins = append(joins, fmt.Sprintf(` LEFT JOIN folder f%d ON f%d.org_id = f%d.org_id AND f%d.uid = f%d.parent_uid`, i, i, i-1, i, i-1))
+	}
+	return strings.Join(joins, "\n")
+}
+
+func getAncestorsSQL(dialect migrator.Dialect, ancestorUIDs []string, start int, end int, origSQL string, origArgs []any) (string, []any) {
```

Comment:
Please confirm that I understand the logic correctly. The code generates the following SQL:

```sql
SELECT f0.id, f0.org_id, f0.uid, f0.parent_uid, f0.title
FROM folder f0
         LEFT JOIN folder f1 ON f1.org_id = f0.org_id AND f1.uid = f0.parent_uid
         LEFT JOIN folder f2 ON f2.org_id = f1.org_id AND f2.uid = f1.parent_uid
         LEFT JOIN folder f3 ON f3.org_id = f2.org_id AND f3.uid = f2.parent_uid
         LEFT JOIN folder f4 ON f4.org_id = f3.org_id AND f4.uid = f3.parent_uid
WHERE f0.org_id = 1
  AND f0.uid IN (<filter>)
  AND (
            f0.uid IN (<permissions>) OR
            f1.uid IN (<permissions>) OR
            f2.uid IN (<permissions>) OR
            f3.uid IN (<permissions>) OR
            f4.uid IN (<permissions>)
    )
```
If I read it correctly, it returns a folder if the user has permission for that folder or any of its ancestors.

For example, 
```
A1
└── B1
    └── C1
		└── D1
    └── C2
        └── D2
A2
└── B2
    └── C3
```
and user requests `D1`, `B2` and `D2` . It has permission to `C1` and `B2`, and therefore, it gets the folder `D1`, `B2` because D1 is a child of C1

My understanding is that the permissions table should have a record for all UIDs on any level of folders the user has access to. So, if the user has permission for `C1` the permission table should contain the record `D1`. If user has access to `A2`, then B2 and C3 should be in the permissions table.

If that's not true, then this logic will have implications on how alerting treats the access to rules and we will have to change this to match (we have permission `alert.rule:read` with scope of folder UID)

Reply:
> My understanding is that the permissions table should have a record for all UIDs on any level of folders the user has access to. So, if the user has permission for C1 the permission table should contain the record D1. If user has access to A2, then B2 and C3 should be in the permissions table.

this is not true: in your example, if in the permissions table there is an entry for C1 then permission for D1 is assumed regardless if there is an entry for D1: D1 `parent_uid` is C1 so this becomes true:
```
AND (
            f0.uid IN (<permissions>) OR
            f1.uid IN (<permissions>) OR
            f2.uid IN (<permissions>) OR
            f3.uid IN (<permissions>) OR
            f4.uid IN (<permissions>)
    )
```

> If that's not true, then this logic will have implications on how alerting treats the access to rules and we will have to change this to match (we have permission alert.rule:read with scope of folder UID)
not sure why: in that case access is given to all rules living under the specific folder UID or its descendants

Reply:
I wonder if it would help if I changed the service to expect any permission action as an additional param to replace the hardcoded [FoldersRead](https://github.com/grafana/grafana/blob/266c071c8c5b2dc9627fe2f3d73dbdf3d74c77b1/pkg/services/folder/folderimpl/folder.go#L131)

Reply:
This is not the only concern. We check user permissions to a path in the web server middleware 
https://github.com/grafana/grafana/blob/6752a512f3908798f948d48b1e607f704f346528/pkg/api/api.go#L439-L458
and 
https://github.com/grafana/grafana/blob/aa03b8f8a7e07054a1747216778712e303520970/pkg/services/ngalert/api/authorization.go#L23-L30

in that places we check that user has permission to a folder by its UID, and I do not see logic similar what is implemented in this PR.

Reply:
I found a code in FolderScopeProvider that checks for ancestors. So, now it's clear. Thanks for explanations!